### PR TITLE
[SPARK-8968] [SQL] external sort by the partition clomns when dynamic partitioning to optimize the memory overhead

### PR DESCRIPTION
**Pull request by scwf**
Now the hash based writer dynamic partitioning show the bad performance for big data and cause many small files and high GC. This patch we do external sort first so that each time we only need open one writer.

before this patch:
gc

after this patch:
gc-optimize-externalsort

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala [hunk 1]
```diff
@@ -115,6 +116,19 @@ case class InsertIntoHiveTable(
...
   }
 
+  def shuffleByDynamicCol(
+      rdd: RDD[InternalRow],
+      writerContainer: SparkHiveWriterContainer): RDD[InternalRow] = writerContainer match {
+    case dwc : SparkHiveDynamicPartitionWriterContainer =>
+      val keys = output.takeRight(dwc.dynamicColNamesLen)
+      val newPartitioning = HashPartitioning(keys, sqlContext.conf.numShufflePartitions)
+      val newRDD = PhysicalRDD(output, rdd)
+      Exchange(newPartitioning, Seq.empty[SortOrder], newRDD).execute()
```
Comment (scwf) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: maybe we can add a config here to control whether to shuffle before insert

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala [hunk 2]
```diff
@@ -115,6 +116,19 @@ case class InsertIntoHiveTable(
     }
   }
 
+  def shuffleByDynamicCol(
```
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: Please make this `private`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala [hunk 3]
```diff
@@ -115,6 +116,19 @@ case class InsertIntoHiveTable(
...
+      rdd: RDD[InternalRow],
+      writerContainer: SparkHiveWriterContainer): RDD[InternalRow] = writerContainer match {
```
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: I'd prefer moving this pattern matching out of this method and put it in `InsertIntoHiveTable`.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala (hunk 1 above)
Comment (liancheng) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala]: Yeah, please add a `SQLConf` option, and probably make it off by default.

**Comment by liancheng** [comment]
I'm thinking maybe we should make the change in `InsertIntoHiveTable.sideEffectResult`, where the writer container gets created. So that you don't bother doing a pattern match later.

Another high level comment is that, although this change does work for your workload, the following statement made in the PR description isn't correct:

> This patch we shuffle data by the partition columns firstly so that each partition will have ony one partition file and this also reduce the gc overhead.

By repartitioning the dataset by dynamic partition columns, you potentially reduce the number of dynamic partitions handled per task (that's why it reduces GC overhead), but the number can't be guaranteed to be reduced to 1.

Actually we are also considering to improve dynamic partitioning insertion via local sorting (sort by partition columns with the spillable `ExternalSorter`). Because when writing sorted data, a task only need to open a single writer, and local sorting doesn't require shuffling.

**Comment by cloud-fan** [comment]
+1 for local sorting. I met OOM when writing parquet with many partitions, and my first idea was also shuffle by partition column. But shuffle is expensive, local sort seems better, we may need to profile these 2 approaches.

**Comment by scwf** [comment]
>This patch we shuffle data by the partition columns firstly so that each partition will have ony one partition file and this also reduce the gc overhead.

@liancheng here i am not mean the partition of spark rdd, i mean for each partition for the table such as `.../a=2/b=5` there will be only one file. before this where maybe many small files under the partition dir.

@cloud-fan @liancheng I have tested this patch and it shows that the performance not become bad(in my situation, it improved 20%-30%).
Anyway i will also have a try to use ExternalSorter to do a local sorting  and will report the performance later.

**Comment by liancheng** [comment]
@scwf Ah I see, these two "partition" concepts are really confusing sometimes... Although I mentioned local sorting, I do tend to also include this repartitioning optimization. But we need to add a `SQLConf` configuration for it first.

**Comment by scwf** [comment]
retest this please

**Comment by liancheng** [comment]
@scwf Could you please help verifying whether the HiveQL CLUSTER BY clause helps in your scenario? Essentially what CLUSTER BY does is just adding an `Exchange` operator, similar to the changes you made in this PR. Sorry that I didn't think of this earlier.

**Comment by rxin** [comment]
Is this related to https://issues.apache.org/jira/browse/SPARK-8890 ?

**Comment by liancheng** [comment]
@rxin Sort of. This PR tries to fix the same issue on the Hive support side.

**Comment by scwf** [comment]
@rxin, yes we found this issue when do dynamic partitioning in our case, so here i do local sort data on the partition columns to reduce the gc overhead.

**Comment by scwf** [comment]
@rxin any comments here?

**Comment by scwf** [comment]
retest this please

**Comment by scwf** [comment]
retest this please

**Comment by marmbrus** [comment]
This looks good, but i'd like to wait until 1.6 since we are now past the deadline and this is a pretty big change.

**Comment by scwf** [comment]
Back to update, @marmbrus @rxin please help review this when you have time.

**Comment by rxin** [comment]
Is this covering a different code path from https://github.com/apache/spark/pull/10638 ?

**Comment by scwf** [comment]
@rxin, yes, This PR try to fix the same issue on the Hive support side.

**Comment by scwf** [comment]
Ping @rxin

**Comment by rxin** [comment]
cc @cloud-fan

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveWriterContainers.scala [hunk 4]
```diff
@@ -198,33 +241,99 @@ private[spark] class SparkHiveDynamicPartitionWriterContainer(
...
+    val standardOI = ObjectInspectorUtils
+      .getStandardObjectInspector(
+        fileSinkConf.getTableInfo.getDeserializer.getObjectInspector,
+        ObjectInspectorCopyOption.JAVA)
+      .asInstanceOf[StructObjectInspector]
+
+    val fieldOIs = standardOI.getAllStructFieldRefs.asScala.map(_.getFieldObjectInspector).toArray
+    val dataTypes = inputSchema.map(_.dataType)
+    val wrappers = fieldOIs.zip(dataTypes).map { case (f, dt) => wrapperFor(f, dt) }
+    val outputData = new Array[Any](fieldOIs.length)
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveWriterContainers.scala]: is the above code duplicated in parent class?

**Comment by cloud-fan** [comment]
retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveWriterContainers.scala (hunk 4 above)
Comment (scwf) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveWriterContainers.scala]: yes, extracted a common method for it.

**Comment by scwf** [comment]
retest this please

**Comment by cloud-fan** [comment]
LGTM

**Comment by rxin** [comment]
I'm going to merge this. The pictures won't really show up in the commit log so it is not that big of a deal, although in the future we should make sure we update it.

**Comment by yhuai** [comment]
@scwf It breaks scala 2.11 build. I am going to fix it.

**Comment by yhuai** [comment]
Fixed by https://github.com/apache/spark/commit/d60f8d74ace5670b1b451a0ea0b93d3b9775bb52.

**Comment by scwf** [comment]
@yhuai  thanks

**Pull request closed without merging**